Author: mmitchel
Date: Sun Dec 11 04:24:42 2005
New Revision: 108376

URL: http://gcc.gnu.org/viewcvs?root=gcc&view=rev&rev=108376
Log:
        PR c++/25010
        * ipa-inline.c (cgraph_clone_inlined_nodes): Do not assume that
        DECL_EXTERNAL functions have no bodies.  Tidy.
        PR c++/25010
        * g++.dg/opt/inline10.C: New test.

--- gcc/cgraphunit.c	2005-12-11 00:58:57.000000000 -0500
+++ gcc/cgraphunit.c	2005-12-11 01:07:57.000000000 -0500
@@ -1070,25 +1070,26 @@
 void
 cgraph_clone_inlined_nodes (struct cgraph_edge *e, bool duplicate)
 {
-  struct cgraph_node *n;
-
-  /* We may eliminate the need for out-of-line copy to be output.  In that
-     case just go ahead and re-use it.  */
-  if (!e->callee->callers->next_caller
-      && (!e->callee->needed || DECL_EXTERNAL (e->callee->decl))
-      && duplicate
-      && flag_unit_at_a_time)
+  if(duplicate)
+  {
+   /* We may eliminate the need for out-of-line copy to be output.
+    * In that case just go ahead and re-use it.  */
+    if (!e->callee->callers->next_caller
+        && !e->callee->needed
+        && flag_unit_at_a_time)
     {
       gcc_assert (!e->callee->global.inlined_to);
-      if (!DECL_EXTERNAL (e->callee->decl))
+      if (DECL_SAVED_TREE (e->callee->decl))
         overall_insns -= e->callee->global.insns, nfunctions_inlined++;
-      duplicate = 0;
+      duplicate = false;
     }
-   else if (duplicate)
+    else
     {
+      struct cgraph_node *n;
       n = cgraph_clone_node (e->callee);
       cgraph_redirect_edge_callee (e, n);
     }
+  }
 
   if (e->caller->global.inlined_to)
     e->callee->global.inlined_to = e->caller->global.inlined_to;

